File: src/main/java/me/lokka30/levelledmobs/managers/NametagQueueManager.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.managers;

import java.time.Instant;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.misc.LivingEntityWrapper;
import me.lokka30.levelledmobs.misc.NametagTimerChecker;
import me.lokka30.levelledmobs.misc.QueueItem;
import me.lokka30.levelledmobs.nametag.NametagSenderHandler;
import me.lokka30.levelledmobs.nametag.NametagSender;
import me.lokka30.levelledmobs.result.NametagResult;
import me.lokka30.levelledmobs.rules.NametagVisibilityEnum;
import me.lokka30.levelledmobs.util.Utils;
import org.bukkit.Bukkit;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.entity.Tameable;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.scheduler.BukkitRunnable;
import org.jetbrains.annotations.NotNull;

/**
 * Queues up mob nametag updates so they can be applied in a background thread
 *
 * @author stumper66
 * @since 3.0.0
 */
public class NametagQueueManager {

    public NametagQueueManager(final LevelledMobs main) {
        this.main = main;
        this.nametagSenderHandler = new NametagSenderHandler(main);
        this.queue = new LinkedBlockingQueue<>();
        getNMSUtil();
    }

    private final LevelledMobs main;
    private boolean isRunning;
    private boolean doThread;
    private NametagSender nametagSender;
    private final LinkedBlockingQueue<QueueItem> queue;
    public final NametagSenderHandler nametagSenderHandler;

    private void getNMSUtil() {

        this.nametagSender = nametagSenderHandler.getCurrentUtil();
    }

    public boolean hasNametagSupport() {
        return this.nametagSender != null;
    }

    public void start() {
        if (isRunning) {
            return;
        }
        doThread = true;
        isRunning = true;

        if (main.getDefinitions().getIsFolia()){
            Consumer<ScheduledTask> bgThread = scheduledTask -> {
                try {
                    mainThread();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                Utils.logger.info("Nametag update queue Manager has exited");
            };

            org.bukkit.Bukkit.getAsyncScheduler().runNow(main, bgThread);
        }
        else{
            final BukkitRunnable bgThread = new BukkitRunnable() {
                @Override
                public void run() {
                    try {
                        mainThread();
                    } catch (final InterruptedException ignored) {
                        isRunning = false;
                    }
                    Utils.logger.info("Nametag update queue Manager has exited");
                }
            };

            bgThread.runTaskAsynchronously(main);
        }
    }

    public void stop() {
        doThread = false;
    }

    void addToQueue(final @NotNull QueueItem item) {
        if (Bukkit.getOnlinePlayers().size() == 0) {
            return;
        }
        if (item.lmEntity.getLivingEntity() == null ||
            main.rulesManager.getRuleCreatureNametagVisbility(item.lmEntity)
                .contains(NametagVisibilityEnum.DISABLED)) {
            return;
        }

        item.lmEntity.inUseCount.getAndIncrement();
        queue.offer(item);
    }

    private void mainThread() throws InterruptedException {
        while (doThread) {

            final QueueItem item = queue.poll(200, TimeUnit.MILLISECONDS);
            if (item == null) {
                continue;
            }

            if (main.getDefinitions().getIsFolia()){
                Consumer<ScheduledTask> task = scheduledTask -> {
                    preProcessItem(item);
                    item.lmEntity.free();
                };

                item.lmEntity.inUseCount.getAndIncrement();
                item.lmEntity.getLivingEntity().getScheduler().run(main, task, null);
            }
            else{
                preProcessItem(item);
            }
        }

        isRunning = false;
    }

    private void preProcessItem(final @NotNull QueueItem item){
        if (item.lmEntity.getLivingEntity() == null) {
            item.lmEntity.free();
            return;
        }

        String lastEntityType = null;
        try {
            lastEntityType = item.lmEntity.getNameIfBaby();
            processItem(item);
        } catch (final Exception ex) {
            final var entityName = lastEntityType == null ? "Unknown Entity" : lastEntityType;

            Utils.logger.error("Unable to process nametag update for '" + entityName + "'. ");
            ex.printStackTrace();
        } finally {
            item.lmEntity.free();
        }
    }

    private void processItem(final @NotNull QueueItem item) {
        if (this.nametagSender == null) {
            // this would happen if the Minecraft version isn't supported directly by NMS
            // and ProtocolLib is not installed
            return;
        }

        final long nametagTimerResetTime = item.lmEntity.getNametagCooldownTime();

        if (nametagTimerResetTime > 0L && !item.nametag.isNullOrEmpty()) {
            synchronized (NametagTimerChecker.nametagTimer_Lock) {
                final Map<Player, WeakHashMap<LivingEntity, Instant>> nametagCooldownQueue = main.nametagTimerChecker.getNametagCooldownQueue();

                if (item.lmEntity.playersNeedingNametagCooldownUpdate != null) {
                    // record which players should get the cooldown for this mob
                    // public Map<Player, WeakHashMap<LivingEntity, Instant>> nametagCooldownQueue;
                    for (final Player player : item.lmEntity.playersNeedingNametagCooldownUpdate) {
                        if (!nametagCooldownQueue.containsKey(player)) {
                            continue;
                        }

                        nametagCooldownQueue.get(player)
                            .put(item.lmEntity.getLivingEntity(), Instant.now());
                        main.nametagTimerChecker.cooldownTimes.put(item.lmEntity.getLivingEntity(),
                            item.lmEntity.getNametagCooldownTime());
                    }

                    // if any players already have a cooldown on this mob then don't remove the cooldown
                    for (final Map.Entry<Player, WeakHashMap<LivingEntity, Instant>> coolDown : nametagCooldownQueue.entrySet()) {
                        final Player player = coolDown.getKey();
                        if (item.lmEntity.playersNeedingNametagCooldownUpdate.contains(player)) {
                            continue;
                        }

                        if (coolDown.getValue().containsKey(item.lmEntity.getLivingEntity())) {
                            item.lmEntity.playersNeedingNametagCooldownUpdate.add(player);
                        }
                    }
                } else {
                    // if there's any existing cooldowns we'll use them
                    for (final Map.Entry<Player, WeakHashMap<LivingEntity, Instant>> coolDown : nametagCooldownQueue.entrySet()) {
                        if (coolDown.getValue().containsKey(item.lmEntity.getLivingEntity())) {
                            if (item.lmEntity.playersNeedingNametagCooldownUpdate == null) {
                                item.lmEntity.playersNeedingNametagCooldownUpdate = new HashSet<>();
                            }

                            item.lmEntity.playersNeedingNametagCooldownUpdate.add(
                                coolDown.getKey());
                        }
                    }
                }
            }
        } else if (item.lmEntity.playersNeedingNametagCooldownUpdate != null) {
            item.lmEntity.playersNeedingNametagCooldownUpdate = null;
        }

        synchronized (NametagTimerChecker.entityTarget_Lock) {
            if (main.nametagTimerChecker.entityTargetMap.containsKey(
                item.lmEntity.getLivingEntity())) {
                if (item.lmEntity.playersNeedingNametagCooldownUpdate == null) {
                    item.lmEntity.playersNeedingNametagCooldownUpdate = new HashSet<>();
                }

                item.lmEntity.playersNeedingNametagCooldownUpdate.add(
                    main.nametagTimerChecker.entityTargetMap.get(item.lmEntity.getLivingEntity()));
            }
        }

        if (!item.lmEntity.getIsPopulated()) {
            return;
        }

        if (main.helperSettings.getBoolean(main.settingsCfg, "use-customname-for-mob-nametags")) {
            updateNametagCustomName(item.lmEntity, item.nametag.getNametag());
            return;
        }

        if (main.helperSettings.getBoolean(main.settingsCfg,
            "assert-entity-validity-with-nametag-packets") && !item.lmEntity.getLivingEntity()
            .isValid()) {
            return;
        }

        updateNametag(item.lmEntity, item.nametag, item.players);
    }

    private void updateNametag(final @NotNull LivingEntityWrapper lmEntity, final @NotNull NametagResult nametag,
        final List<Player> players) {
        final int loopCount = lmEntity.playersNeedingNametagCooldownUpdate == null ?
            1 : 2;

        for (int i = 0; i < loopCount; i++) {
            // will loop again to update with nametag cooldown for only the specified players

            final List<NametagVisibilityEnum> nametagVisibilityEnum = main.rulesManager.getRuleCreatureNametagVisbility(
                lmEntity);
            final boolean doAlwaysVisible = i == 1 ||
                    !nametag.isNullOrEmpty() && lmEntity.getLivingEntity().isCustomNameVisible() ||
                nametagVisibilityEnum.contains(NametagVisibilityEnum.ALWAYS_ON);

            if (i == 0) {
                // these players are not getting always on nametags unless always on has been configured for the mob
                for (final Player player : players) {
                    if (lmEntity.playersNeedingNametagCooldownUpdate != null
                        && lmEntity.playersNeedingNametagCooldownUpdate.contains(player)) {
                        continue;
                    }

                    nametagSender.sendNametag(lmEntity.getLivingEntity(), nametag, player,
                        doAlwaysVisible);
                }
            } else {
                // these players are getting always on nametags
                for (final Player player : lmEntity.playersNeedingNametagCooldownUpdate) {
                    nametagSender.sendNametag(lmEntity.getLivingEntity(), nametag, player, true);
                }
            }
        }
    }

    @SuppressWarnings("deprecation")
    private void updateNametagCustomName(final @NotNull LivingEntityWrapper lmEntity,
                                         final String nametag) {
        synchronized (lmEntity.getLivingEntity().getPersistentDataContainer()) {
            if (lmEntity.getPDC()
                .has(main.namespacedKeys.hasCustomNameTag, PersistentDataType.INTEGER)) {
                return;
            }
        }

        final boolean hadCustomName = lmEntity.getLivingEntity().getCustomName() != null;

        lmEntity.getLivingEntity().setCustomName(nametag);
        lmEntity.getLivingEntity().setCustomNameVisible(true);

        final boolean isTamable = (lmEntity.getLivingEntity() instanceof Tameable);

        if (!hadCustomName && !isTamable && !lmEntity.getTypeName().equalsIgnoreCase("Axolotl")) {
            lmEntity.getLivingEntity().setRemoveWhenFarAway(true);
        }
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/managers/NametagQueueManager.kt
package io.github.arcaneplugins.levelledmobs.managers

import java.time.Instant
import java.util.WeakHashMap
import java.util.concurrent.LinkedBlockingQueue
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.misc.NametagTimerChecker
import io.github.arcaneplugins.levelledmobs.misc.QueueItem
import io.github.arcaneplugins.levelledmobs.nametag.NametagSender
import io.github.arcaneplugins.levelledmobs.nametag.NametagSenderHandler
import io.github.arcaneplugins.levelledmobs.result.NametagResult
import io.github.arcaneplugins.levelledmobs.enums.NametagVisibilityEnum
import io.github.arcaneplugins.levelledmobs.util.LibsDisguisesUtils
import io.github.arcaneplugins.levelledmobs.util.Log
import io.github.arcaneplugins.levelledmobs.util.MessageUtils
import io.github.arcaneplugins.levelledmobs.util.Utils
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import io.github.arcaneplugins.levelledmobs.wrappers.SchedulerWrapper
import org.bukkit.Bukkit
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitTask

/**
 * Queues up mob nametag updates so they can be applied in a background thread
 *
 * @author stumper66
 * @since 3.0.0
 */
class NametagQueueManager {
    private var isRunning = false
    private var doThread = false
    private var nametagSender: NametagSender? = null
    private var hasLibsDisguisesInstalled = false
    var disableNametagJava = false
    var disableNametagBedrock = false
    var queueTask: BukkitTask? = null
    private val queue = LinkedBlockingQueue<QueueItem>()
    val nametagSenderHandler = NametagSenderHandler()
    private val queueLock = Any()

    fun load(){
        hasLibsDisguisesInstalled = ExternalCompatibilityManager.hasLibsDisguisesInstalled
        this.nametagSender = nametagSenderHandler.getCurrentUtil()
        onLoadOrReload()
    }

    fun onLoadOrReload(){
        this.disableNametagJava = LevelledMobs.instance.helperSettings.getBoolean(
            "disable-nametag-java", false
        )
        this.disableNametagBedrock = LevelledMobs.instance.helperSettings.getBoolean(
            "disable-nametag-bedrock", false
        )
    }

    val hasNametagSupport: Boolean
        get() = this.nametagSender != null

    fun start() {
        // folia will run directly
        if (LevelledMobs.instance.ver.isRunningFolia) return

        if (isRunning) return

        doThread = true
        isRunning = true

        val scheduler = SchedulerWrapper {
            var hadError = false
            try {
                mainThread()
            } catch (e: Exception) {
                if (e !is InterruptedException){
                    hadError = true
                    e.printStackTrace()
                }
            }
            if (hadError)
                Log.sev("Nametag update queue Manager has exited with error")
            else
                Log.inf("Nametag update queue Manager has exited")

            isRunning = false
        }
        scheduler.run()
        this.queueTask = scheduler.bukkitTask
    }

    fun stop() {
        doThread = false
    }

    fun taskChecker(){
        val qt = queueTask ?: return

        val queueSize = getNumberQueued()

        if (queueSize < 1000 && !qt.isCancelled || Bukkit.getScheduler().isCurrentlyRunning(qt.taskId)) return
        val status = if (qt.isCancelled) "cancelled"
        else if (queueSize < 1000) "not running"
        else "queue size was $queueSize"

        Log.war("Restarting Nametag Queue Manager task, status was $status")
        qt.cancel()
        isRunning = false
        start()
    }

    fun addToQueue(item: QueueItem) {
        if (!item.lmEntity.shouldShowLMNametag) return
        if (Bukkit.getOnlinePlayers().isEmpty()) return
        if (item.lmEntity.nametagVisibilityEnum.contains(NametagVisibilityEnum.DISABLED))
            return

        item.lmEntity.inUseCount.getAndIncrement()

        if (LevelledMobs.instance.ver.isRunningFolia){
            // folia runs directly
            preProcessItem(item)
        }
        else{
            synchronized(queueLock){
                queue.offer(item)
            }
        }
    }

    fun getNumberQueued(): Int{
        val size: Int
        synchronized(queueLock){
            size = queue.size
        }

        return size
    }

    private fun mainThread() {
        while (doThread) {
            val item: QueueItem?

            synchronized(queueLock){
                item = queue.poll()
            }
            if (item == null) {
                Thread.sleep(2L)
                continue
            }

            val scheduler = SchedulerWrapper(
                item.lmEntity.livingEntity
            ) {
                preProcessItem(item)
                item.lmEntity.free()
            }

            scheduler.runDirectlyInBukkit = true
            scheduler.entity = item.lmEntity.livingEntity
            item.lmEntity.inUseCount.getAndIncrement()
            scheduler.run()
        }

        isRunning = false
    }

    private fun preProcessItem(item: QueueItem) {
        if (!item.lmEntity.isPopulated) {
            item.lmEntity.free()
            return
        }

        var lastEntityType: String? = null
        try {
            lastEntityType = item.lmEntity.nameIfBaby
            processItem(item)
        } catch (ex: Exception) {
            val entityName = lastEntityType ?: "Unknown Entity"

            Log.sev("Unable to process nametag update for '$entityName'. ")
            ex.printStackTrace()
        } finally {
            item.lmEntity.free()
        }
    }

    private fun processItem(item: QueueItem) {
        if (this.nametagSender == null) {
            // this would happen if the Minecraft version isn't supported directly by NMS
            return
        }

        val nametagTimerResetTime = item.lmEntity.getNametagCooldownTime()

        if (nametagTimerResetTime > 0L && !item.nametag!!.isNullOrEmpty) {
            synchronized(NametagTimerChecker.nametagTimer_Lock) {
                val nametagCooldownQueue: Map<Player, WeakHashMap<LivingEntity, Instant>> =
                    LevelledMobs.instance.nametagTimerChecker.nametagCooldownQueue
                if (item.lmEntity.playersNeedingNametagCooldownUpdate != null) {
                    // record which players should get the cooldown for this mob
                    // public Map<Player, WeakHashMap<LivingEntity, Instant>> nametagCooldownQueue;
                    for (player in item.lmEntity.playersNeedingNametagCooldownUpdate!!) {
                        if (!nametagCooldownQueue.containsKey(player)) {
                            continue
                        }

                        nametagCooldownQueue[player]?.set(item.lmEntity.livingEntity, Instant.now())
                        LevelledMobs.instance.nametagTimerChecker.cooldownTimes[item.lmEntity.livingEntity] =
                            item.lmEntity.getNametagCooldownTime()
                    }

                    // if any players already have a cooldown on this mob then don't remove the cooldown
                    for ((player, value) in nametagCooldownQueue) {
                        if (item.lmEntity.playersNeedingNametagCooldownUpdate!!.contains(player)) {
                            continue
                        }

                        if (value.containsKey(item.lmEntity.livingEntity)) {
                            item.lmEntity.playersNeedingNametagCooldownUpdate!!.add(player)
                        }
                    }
                } else {
                    // if there's any existing cooldowns we'll use them
                    for ((key, value) in nametagCooldownQueue) {
                        if (value.containsKey(item.lmEntity.livingEntity)) {
                            if (item.lmEntity.playersNeedingNametagCooldownUpdate == null) {
                                item.lmEntity.playersNeedingNametagCooldownUpdate = HashSet()
                            }

                            item.lmEntity.playersNeedingNametagCooldownUpdate!!.add(
                                key
                            )
                        }
                    }
                }
            }
        } else if (item.lmEntity.playersNeedingNametagCooldownUpdate != null) {
            item.lmEntity.playersNeedingNametagCooldownUpdate = null
        }

        val main = LevelledMobs.instance
        synchronized(NametagTimerChecker.entityTarget_Lock) {
            if (main.nametagTimerChecker.entityTargetMap.containsKey(
                    item.lmEntity.livingEntity
                )
            ) {
                if (item.lmEntity.playersNeedingNametagCooldownUpdate == null) {
                    item.lmEntity.playersNeedingNametagCooldownUpdate = mutableSetOf()
                }

                item.lmEntity.playersNeedingNametagCooldownUpdate!!.add(
                    main.nametagTimerChecker.entityTargetMap[item.lmEntity.livingEntity]!!
                )
            }
        }

        if (!item.lmEntity.isPopulated) {
            return
        }

        if (main.helperSettings.getBoolean(
                "assert-entity-validity-with-nametag-packets"
            ) && !item.lmEntity.livingEntity
                .isValid
        ) {
            return
        }

        updateNametag(item.lmEntity, item.nametag!!, item.players!!)
    }

    @Suppress("DEPRECATION")
    private fun updateNametag(
        lmEntity: LivingEntityWrapper,
        nametag: NametagResult,
        players: MutableList<Player>
    ) {
        val loopCount = if (lmEntity.playersNeedingNametagCooldownUpdate == null) 1 else 2

        for (i in 0 until loopCount) {
            // will loop again to update with nametag cooldown for only the specified players

            val nametagVisibilityEnum = lmEntity.nametagVisibilityEnum
            val doAlwaysVisible = i == 1 || !nametag.isNullOrEmpty && lmEntity.livingEntity.isCustomNameVisible ||
                    nametagVisibilityEnum.contains(NametagVisibilityEnum.ALWAYS_ON)

            if (i == 0) {
                // these players are not getting always on nametags unless always on has been configured for the mob
                for (player in players) {
                    if (lmEntity.playersNeedingNametagCooldownUpdate != null
                        && lmEntity.playersNeedingNametagCooldownUpdate!!.contains(player)
                    ) {
                        continue
                    }
                    
                    /** Disable if Java or Bedrock */
                    if (disableNametagBedrock && isBedrock(player)) return
                    if (disableNametagJava && !isBedrock(player)) return

                    nametagSender!!.sendNametag(
                        lmEntity.livingEntity, nametag, player,
                        doAlwaysVisible
                    )
                }
            } else {
                // these players are getting always on nametags
                for (player in lmEntity.playersNeedingNametagCooldownUpdate!!) {

                    /** Disable if Java or Bedrock */
                    if (disableNametagBedrock && isBedrock(player)) return
                    if (disableNametagJava && !isBedrock(player)) return


                    nametagSender!!.sendNametag(lmEntity.livingEntity, nametag, player, true)
                }
            }

            if (hasLibsDisguisesInstalled && LibsDisguisesUtils.isMobUsingLibsDisguises(lmEntity)) {
                var useNametag: String? = null
                if (nametag.nametag != null){
                    useNametag = MessageUtils.colorizeAll(nametag.nametagNonNull
                        .replace("{DisplayName}", Utils.capitalize(lmEntity.typeName.replace("_", " ")))
                        .replace("{CustomName}", lmEntity.livingEntity.customName ?: ""))
                }

                LibsDisguisesUtils.updateLibsDisguiseNametag(lmEntity, useNametag)
            }
        }
    }

    private fun isBedrock(player: Player) : Boolean {
        return player.uniqueId.mostSignificantBits == 0L
    }
}